### PR TITLE
[SPARK-33758][SQL] Prune unrequired partitionings from AliasAwareOutputPartitionings when some columns are dropped from projection 

### PR DESCRIPTION
**Pull request by prakharjain09**
### What changes were proposed in this pull request?
This PR tries to prune the unrequired output partitionings in cases when the columns are dropped from Project/Aggregates etc.

### Why are the changes needed?
Consider this query:
    select t1.id from t1 JOIN t2 on t1.id = t2.id

This query will have top level Project node which will just project t1.id. But the outputPartitioning of this project node will be: PartitioningCollection(HashPartitioning(t1.id), HashPartitioning(t2.id)).

But since we are not propagating t2.id column, so we can drop HashPartitioning(t2.id) from the output partitioning of Project node.

 
### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UTs.

### CODE REVIEW

**Comment by prakharjain09** [comment]
cc - @maropu @cloud-fan

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
add to whitelist

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 1]
```diff
@@ -1008,17 +1008,11 @@ class PlannerSuite extends SharedSparkSession with AdaptiveSparkPlanHelper {
...
         assert(projects.exists(_.outputPartitioning match {
-          case PartitioningCollection(Seq(HashPartitioning(Seq(Multiply(ar1, _, _)), _),
-            HashPartitioning(Seq(Multiply(ar2, _, _)), _))) =>
-            Seq(ar1, ar2) match {
-              case Seq(ar1: AttributeReference, ar2: AttributeReference) =>
-                ar1.name == "t1id" && ar2.name == "id2"
-              case _ =>
-                false
-            }
+          case HashPartitioning(Seq(Multiply(ar1: AttributeReference, _, _)), _)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:1011]: nit:
```
case ... =>
  ar1.name == "t1id"
```

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala (hunk 1 above)
Comment (prakharjain09) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: done.

**Comment by prakharjain09** [comment]
Thanks @cloud-fan for the review. I have addressed your comments.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by cloud-fan** [comment]
BTW, shall we do the same to output ordering? We can prune `SortOrder.sameOrderExpressions`

**Comment by prakharjain09** [comment]
> BTW, shall we do the same to output ordering? We can prune SortOrder.sameOrderExpressions

@cloud-fan Sure. WIll make changes for it and raise PR in couple of days.

**Comment by maropu** [comment]
late lgtm. Thanks for fixing it, @prakharjain09

**Pull request closed without merging**